File: multi-auto-agent/multi_auto_agent/systems/worktree.py
import json
import re
import subprocess
import time
from pathlib import Path
from ..config import REPO_ROOT

class EventBus:
    def __init__(self, event_log_path: Path):
        self.path = event_log_path
        self.path.parent.mkdir(parents=True, exist_ok=True)
        if not self.path.exists():
            self.path.write_text("")

    def emit(self, event: str, task: dict | None = None, worktree: dict | None = None, error: str | None = None):
        payload = {
            "event": event,
            "ts": time.time(),
            "task": task or {},
            "worktree": worktree or {},
        }
        if error:
            payload["error"] = error
        with self.path.open("a", encoding="utf-8") as f:
            f.write(json.dumps(payload) + "\n")

    def list_recent(self, limit: int = 20) -> str:
        n = max(1, min(int(limit or 20), 200))
        lines = self.path.read_text(encoding="utf-8").splitlines()
        recent = lines[-n:]
        items = []
        for line in recent:
            try:
                items.append(json.loads(line))
            except Exception:
                items.append({"event": "parse_error", "raw": line})
        return json.dumps(items, indent=2)

class WorktreeManager:
    def __init__(self, repo_root: Path, tasks, events: EventBus):
        self.repo_root = repo_root
        self.tasks = tasks
        self.events = events
        self.dir = repo_root / ".worktrees"
        self.dir.mkdir(parents=True, exist_ok=True)
        self.index_path = self.dir / "index.json"
        if not self.index_path.exists():
            self.index_path.write_text(json.dumps({"worktrees": []}, indent=2))
        self.git_available = self._is_git_repo()

    def _is_git_repo(self) -> bool:
        try:
            r = subprocess.run(
                ["git", "rev-parse", "--is-inside-work-tree"],
                cwd=self.repo_root,
                capture_output=True,
                text=True,
                timeout=10,
            )
            return r.returncode == 0
        except Exception:
            return False

    def _run_git(self, args: list[str]) -> str:
        if not self.git_available:
            raise RuntimeError("Not in a git repository. worktree tools require git.")
        r = subprocess.run(
            ["git", *args],
            cwd=self.repo_root,
            capture_output=True,
            text=True,
            timeout=120,
        )
        if r.returncode != 0:
            msg = (r.stdout + r.stderr).strip()
            raise RuntimeError(msg or f"git {' '.join(args)} failed")
        return (r.stdout + r.stderr).strip() or "(no output)"

    def _load_index(self) -> dict:
        return json.loads(self.index_path.read_text())

    def _save_index(self, data: dict):
        self.index_path.write_text(json.dumps(data, indent=2))

    def _find(self, name: str) -> dict | None:
        idx = self._load_index()
        for wt in idx.get("worktrees", []):
            if wt.get("name") == name:
                return wt
        return None

    def _validate_name(self, name: str):
        if not re.fullmatch(r"[A-Za-z0-9._-]{1,40}", name or ""):
            raise ValueError("Invalid worktree name. Use 1-40 chars: letters, numbers, ., _, -")

    def create(self, name: str, task_id: int = None, base_ref: str = "HEAD") -> str:
        self._validate_name(name)
        if self._find(name):
            raise ValueError(f"Worktree '{name}' already exists in index")
        if task_id is not None and not self.tasks.exists(task_id):
            raise ValueError(f"Task {task_id} not found")

        path = self.dir / name
        branch = f"wt/{name}"
        self.events.emit("worktree.create.before", task={"id": task_id} if task_id is not None else {}, worktree={"name": name, "base_ref": base_ref})
        try:
            self._run_git(["worktree", "add", "-b", branch, str(path), base_ref])

            entry = {
                "name": name,
                "path": str(path),
                "branch": branch,
                "task_id": task_id,
                "status": "active",
                "created_at": time.time(),
            }

            idx = self._load_index()
            idx["worktrees"].append(entry)
            self._save_index(idx)

            if task_id is not None:
                self.tasks.bind_worktree(task_id, name)

            self.events.emit("worktree.create.after", task={"id": task_id} if task_id is not None else {}, worktree={"name": name, "path": str(path), "branch": branch, "status": "active"})
            return json.dumps(entry, indent=2)
        except Exception as e:
            self.events.emit("worktree.create.failed", task={"id": task_id} if task_id is not None else {}, worktree={"name": name, "base_ref": base_ref}, error=str(e))
            raise

    def list_all(self) -> str:
        idx = self._load_index()
        wts = idx.get("worktrees", [])
        if not wts:
            return "No worktrees in index."
        lines = []
        for wt in wts:
            suffix = f" task={wt['task_id']}" if wt.get("task_id") else ""
            lines.append(f"[{wt.get('status', 'unknown')}] {wt['name']} -> {wt['path']} ({wt.get('branch', '-')}){suffix}")
        return "\n".join(lines)

    def status(self, name: str) -> str:
        wt = self._find(name)
        if not wt: return f"Error: Unknown worktree '{name}'"
        path = Path(wt["path"])
        if not path.exists(): return f"Error: Worktree path missing: {path}"
        r = subprocess.run(["git", "status", "--short", "--branch"], cwd=path, capture_output=True, text=True, timeout=60)
        return (r.stdout + r.stderr).strip() or "Clean worktree"

    def run(self, name: str, command: str) -> str:
        dangerous = ["rm -rf /", "sudo", "shutdown", "reboot", "> /dev/"]
        if any(d in command for d in dangerous):
            return "Error: Dangerous command blocked"

        wt = self._find(name)
        if not wt: return f"Error: Unknown worktree '{name}'"
        path = Path(wt["path"])
        if not path.exists(): return f"Error: Worktree path missing: {path}"

        try:
            r = subprocess.run(command, shell=True, cwd=path, capture_output=True, text=True, timeout=300)
            out = (r.stdout + r.stderr).strip()
            return out[:50000] if out else "(no output)"
        except subprocess.TimeoutExpired:
            return "Error: Timeout (300s)"

    def remove(self, name: str, force: bool = False, complete_task: bool = False) -> str:
        wt = self._find(name)
        if not wt: return f"Error: Unknown worktree '{name}'"

        self.events.emit("worktree.remove.before", task={"id": wt.get("task_id")} if wt.get("task_id") is not None else {}, worktree={"name": name, "path": wt.get("path")})
        try:
            args = ["worktree", "remove"]
            if force: args.append("--force")
            args.append(wt["path"])
            self._run_git(args)

            if complete_task and wt.get("task_id") is not None:
                task_id = wt["task_id"]
                before = json.loads(self.tasks.get(task_id))
                self.tasks.update(task_id, status="completed")
                self.tasks.unbind_worktree(task_id)
                self.events.emit("task.completed", task={"id": task_id, "subject": before.get("subject", ""), "status": "completed"}, worktree={"name": name})

            idx = self._load_index()
            for item in idx.get("worktrees", []):
                if item.get("name") == name:
                    item["status"] = "removed"
                    item["removed_at"] = time.time()
            self._save_index(idx)

            self.events.emit("worktree.remove.after", task={"id": wt.get("task_id")} if wt.get("task_id") is not None else {}, worktree={"name": name, "path": wt.get("path"), "status": "removed"})
            return f"Removed worktree '{name}'"
        except Exception as e:
            self.events.emit("worktree.remove.failed", task={"id": wt.get("task_id")} if wt.get("task_id") is not None else {}, worktree={"name": name, "path": wt.get("path")}, error=str(e))
            raise

    def keep(self, name: str) -> str:
        wt = self._find(name)
        if not wt: return f"Error: Unknown worktree '{name}'"

        idx = self._load_index()
        kept = None
        for item in idx.get("worktrees", []):
            if item.get("name") == name:
                item["status"] = "kept"
                item["kept_at"] = time.time()
                kept = item
        self._save_index(idx)

        self.events.emit("worktree.keep", task={"id": wt.get("task_id")} if wt.get("task_id") is not None else {}, worktree={"name": name, "path": wt.get("path"), "status": "kept"})
        return json.dumps(kept, indent=2) if kept else f"Error: Unknown worktree '{name}'"

File: src/application/services/worktree.ts
import fs from 'fs';
import path from 'path';
import { exec } from 'child_process';
import { WORKDIR } from '../../infra/config';
import { TaskManager } from './task-manager';
import { promisify } from 'util';

const execAsync = promisify(exec);

export interface Worktree {
  name: string;
  path: string;
  branch: string;
  task_id: number | null;
  status: 'active' | 'removed' | 'kept';
  created_at: number;
  removed_at?: number;
  kept_at?: number;
}

export interface WorktreeIndex {
  worktrees: Worktree[];
}

export class EventBus {
  private path: string;

  constructor(eventLogPath: string) {
    this.path = eventLogPath;
    if (!fs.existsSync(path.dirname(this.path))) {
      fs.mkdirSync(path.dirname(this.path), { recursive: true });
    }
    if (!fs.existsSync(this.path)) {
      fs.writeFileSync(this.path, '');
    }
  }

  emit(event: string, task: any = {}, worktree: any = {}, error?: string): void {
    const payload = {
      event,
      ts: Date.now() / 1000,
      task: task || {},
      worktree: worktree || {},
      ...(error ? { error } : {})
    };
    fs.appendFileSync(this.path, JSON.stringify(payload) + '\n');
  }

  listRecent(limit: number = 20): string {
    const n = Math.max(1, Math.min(limit || 20, 200));
    const content = fs.readFileSync(this.path, 'utf-8');
    const lines = content.split('\n').filter(Boolean);
    const recent = lines.slice(-n);
    const items = recent.map(line => {
      try {
        return JSON.parse(line);
      } catch {
        return { event: 'parse_error', raw: line };
      }
    });
    return JSON.stringify(items, null, 2);
  }
}

export class WorktreeManager {
  private repoRoot: string;
  private tasks: TaskManager;
  private events: EventBus;
  private dir: string;
  private indexPath: string;
  public gitAvailable: boolean = false;

  constructor(repoRoot: string, tasks: TaskManager, events: EventBus) {
    this.repoRoot = repoRoot;
    this.tasks = tasks;
    this.events = events;
    this.dir = path.join(repoRoot, '..', `${path.basename(repoRoot)}-worktrees`);
    this.indexPath = path.join(this.dir, 'index.json');

    if (!fs.existsSync(this.dir)) {
      fs.mkdirSync(this.dir, { recursive: true });
    }
    if (!fs.existsSync(this.indexPath)) {
      fs.writeFileSync(this.indexPath, JSON.stringify({ worktrees: [] }, null, 2));
    }

    this.checkGitRepo();
  }

  private async checkGitRepo() {
    try {
      await execAsync('git rev-parse --is-inside-work-tree', { cwd: this.repoRoot });
      this.gitAvailable = true;
    } catch {
      this.gitAvailable = false;
    }
  }

  private async runGit(args: string[]): Promise<string> {
    if (!this.gitAvailable) {
      throw new Error("Not in a git repository. worktree tools require git.");
    }
    try {
      const { stdout, stderr } = await execAsync(`git ${args.join(' ')}`, { cwd: this.repoRoot, timeout: 120000 });
      return (stdout + stderr).trim() || "(no output)";
    } catch (error: any) {
      const msg = (error.stdout + error.stderr).trim();
      throw new Error(msg || `git ${args.join(' ')} failed`);
    }
  }

  private loadIndex(): WorktreeIndex {
    return JSON.parse(fs.readFileSync(this.indexPath, 'utf-8'));
  }

  private saveIndex(data: WorktreeIndex): void {
    fs.writeFileSync(this.indexPath, JSON.stringify(data, null, 2));
  }

  private find(name: string): Worktree | undefined {
    const idx = this.loadIndex();
    return idx.worktrees.find(wt => wt.name === name);
  }

  private validateName(name: string): void {
    if (!/^[A-Za-z0-9._-]{1,40}$/.test(name || "")) {
      throw new Error("Invalid worktree name. Use 1-40 chars: letters, numbers, ., _, -");
    }
  }

  async create(name: string, taskId?: number, baseRef: string = "HEAD"): Promise<string> {
    this.validateName(name);
    if (this.find(name)) {
      throw new Error(`Worktree '${name}' already exists in index`);
    }
    if (taskId !== undefined) {
      // Check if task exists (we'll implement exist check in TaskManager)
      try {
        this.tasks.get(taskId);
      } catch {
        throw new Error(`Task ${taskId} not found`);
      }
    }

    const wtPath = path.join(this.dir, name);
    const branch = `wt/${name}`;
    this.events.emit("worktree.create.before", taskId !== undefined ? { id: taskId } : {}, { name, base_ref: baseRef });

    try {
      await this.runGit(["worktree", "add", "-b", branch, wtPath, baseRef]);

      const entry: Worktree = {
        name,
        path: wtPath,
        branch,
        task_id: taskId ?? null,
        status: 'active',
        created_at: Date.now() / 1000
      };

      const idx = this.loadIndex();
      idx.worktrees.push(entry);
      this.saveIndex(idx);

      if (taskId !== undefined) {
        this.tasks.bindWorktree(taskId, name);
      }

      this.events.emit("worktree.create.after", taskId !== undefined ? { id: taskId } : {}, { name, path: wtPath, branch, status: 'active' });
      return JSON.stringify(entry, null, 2);
    } catch (e: any) {
      this.events.emit("worktree.create.failed", taskId !== undefined ? { id: taskId } : {}, { name, base_ref: baseRef }, e.message);
      throw e;
    }
  }

  listAll(): string {
    const idx = this.loadIndex();
    const wts = idx.worktrees;
    if (wts.length === 0) return "No worktrees in index.";
    
    return wts.map(wt => {
      const suffix = wt.task_id ? ` task=${wt.task_id}` : "";
      return `[${wt.status}] ${wt.name} -> ${wt.path} (${wt.branch})${suffix}`;
    }).join('\n');
  }

  async status(name: string): Promise<string> {
    const wt = this.find(name);
    if (!wt) return `Error: Unknown worktree '${name}'`;
    
    if (!fs.existsSync(wt.path)) return `Error: Worktree path missing: ${wt.path}`;

    try {
      const { stdout, stderr } = await execAsync("git status --short --branch", { cwd: wt.path, timeout: 60000 });
      return (stdout + stderr).trim() || "Clean worktree";
    } catch (e: any) {
      return (e.stdout + e.stderr).trim();
    }
  }

  async run(name: string, command: string): Promise<string> {
    const dangerous = ["rm -rf /", "sudo", "shutdown", "reboot", "> /dev/"];
    if (dangerous.some(d => command.includes(d))) {
      return "Error: Dangerous command blocked";
    }

    const wt = this.find(name);
    if (!wt) return `Error: Unknown worktree '${name}'`;
    if (!fs.existsSync(wt.path)) return `Error: Worktree path missing: ${wt.path}`;

    try {
      const { stdout, stderr } = await execAsync(command, { cwd: wt.path, timeout: 300000 });
      const out = (stdout + stderr).trim();
      return out.length > 50000 ? out.substring(0, 50000) : (out || "(no output)");
    } catch (e: any) {
       if (e.code === 'ETIMEDOUT') return "Error: Timeout (300s)";
       const msg = (e.stdout + e.stderr).trim();
       return msg || `Error: ${e.message}`;
    }
  }

  async remove(name: string, force: boolean = false, completeTask: boolean = false): Promise<string> {
    const wt = this.find(name);
    if (!wt) return `Error: Unknown worktree '${name}'`;

    this.events.emit("worktree.remove.before", wt.task_id !== null ? { id: wt.task_id } : {}, { name, path: wt.path });

    try {
      const args = ["worktree", "remove"];
      if (force) args.push("--force");
      args.push(wt.path);
      await this.runGit(args);

      if (completeTask && wt.task_id !== null) {
        const taskId = wt.task_id;
        const taskJson = this.tasks.get(taskId);
        const before = JSON.parse(taskJson);
        this.tasks.update(taskId, "completed");
        this.tasks.unbindWorktree(taskId);
        this.events.emit("task.completed", { id: taskId, subject: before.subject, status: "completed" }, { name });
      }

      const idx = this.loadIndex();
      const item = idx.worktrees.find(w => w.name === name);
      if (item) {
        item.status = 'removed';
        item.removed_at = Date.now() / 1000;
      }
      this.saveIndex(idx);

      this.events.emit("worktree.remove.after", wt.task_id !== null ? { id: wt.task_id } : {}, { name, path: wt.path, status: 'removed' });
      return `Removed worktree '${name}'`;
    } catch (e: any) {
      this.events.emit("worktree.remove.failed", wt.task_id !== null ? { id: wt.task_id } : {}, { name, path: wt.path }, e.message);
      throw e;
    }
  }

  keep(name: string): string {
    const wt = this.find(name);
    if (!wt) return `Error: Unknown worktree '${name}'`;

    const idx = this.loadIndex();
    const item = idx.worktrees.find(w => w.name === name);
    if (item) {
      item.status = 'kept';
      item.kept_at = Date.now() / 1000;
    }
    this.saveIndex(idx);
    
    const kept = idx.worktrees.find(w => w.name === name);

    this.events.emit("worktree.keep", wt.task_id !== null ? { id: wt.task_id } : {}, { name, path: wt.path, status: 'kept' });
    return JSON.stringify(kept, null, 2);
  }
}
